x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupAction.java

```diff
@@ -288,47 +289,61 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
-                WriteableZoneId rollupDateZoneId = WriteableZoneId.of(dateConfig.getTimeZone());
-                if (rollupGroups.containsKey(rollupGroupKeyName)) {
-                    RollupGroup group = rollupGroups.get(rollupGroupKeyName);
-                    group.add(rollupIndexName, dateConfig.getInterval(), rollupDateZoneId);
-                } else {
-                    RollupGroup group = new RollupGroup();
-                    group.add(rollupIndexName, dateConfig.getInterval(), rollupDateZoneId);
-                    rollupGroups.put(rollupGroupKeyName, group);
-                }
-                // add rolled up index to backing datastream if rolling up a backing index of a datastream
-                IndexAbstraction originalIndex = currentState.getMetadata().getIndicesLookup().get(originalIndexName);
+
                 DataStream dataStream = null;
                 if (originalIndex.getParentDataStream() != null) {
+                    // If rolling up a backing index of a datastream, add rolled up index to backing datastream
                     DataStream originalDataStream = originalIndex.getParentDataStream().getDataStream();
+
+                    Map<String, Object> dsMetadata = originalDataStream.getMetadata() != null
+                        ? originalDataStream.getMetadata() : new HashMap<>();
+                    final RollupMetadata rollupMetadata = dsMetadata.containsKey(RollupMetadata.TYPE) ?
+                        (RollupMetadata) dsMetadata.get(RollupMetadata.TYPE) : null;
+                    final Map<String, RollupGroup> rollupGroups = rollupMetadata != null ?
+                        new HashMap<>(rollupMetadata.rollupGroups()) : new HashMap<>();
+
+                    RollupActionConfig rollupConfig = request.getRollupConfig();
+                    RollupActionDateHistogramGroupConfig dateConfig = rollupConfig.getGroupConfig().getDateHistogram();
+                    WriteableZoneId rollupDateZoneId = WriteableZoneId.of(dateConfig.getTimeZone());
+                    Map<String, List<String>> metricsConfig = new HashMap<>();
+                    for (MetricConfig mconfig: rollupConfig.getMetricsConfig()) {
+                        metricsConfig.put(mconfig.getField(), mconfig.getMetrics());
+                    }
+                    RollupIndexMetadata rollupInfo = new RollupIndexMetadata(dateConfig.getInterval(), rollupDateZoneId, metricsConfig);
+
+                    if (rollupGroups.containsKey(rollupGroupKeyName)) {
+                        RollupGroup group = rollupGroups.get(rollupGroupKeyName);
+                        group.add(rollupIndexName, rollupInfo);
+                    } else {
+                        RollupGroup group = new RollupGroup();
+                        group.add(rollupIndexName, rollupInfo);
+                        rollupGroups.put(rollupGroupKeyName, group);
+                    }
+                    dsMetadata.put(RollupMetadata.TYPE, new RollupMetadata(rollupGroups));
```

Comment:
Since this metadata is intended for metadata defined in composable templates (#63991), I think we should have RollupMetadata be its own field within DataStream.

Reply:
Thanks for the comment. We also discussed that this will probably solve the problem we faced when serializing the `RollupMetadata` object when added in the `_meta` map. 

However, I am still not sure if it's ok to make the datastream aware of rollups. 
Any feedback from the @elastic/es-core-features team would be welcome

Reply:
On another note, adding the `RollupMetadata` instance inside the `DataStream` object makes it almost impossible to guard against the RollupsV2 feature flag

Reply:
good point. the one upside is that it will never be non-null when rollupV2 is not used. So we do not have to worry about existing usage of it since this is internal and we have the assignment of this instance variable under the RollupV2 feature flag.

Reply:
I pushed implementation of this in a separate branch.

You can view it here:
https://github.com/csoulios/elasticsearch/compare/rollup-metadata-ds..csoulios:rollup-metadata-ds-own-field?diff=split

Reply:
I prefer this approach that you've laid out in `rollup-metadata-ds-own-field`. I would just say that `NEW_FEATURES_VERSION` should not be used to version-guard the new field, but overall, yes I think it is good. Maybe the DataStream team has differing opinion, not sure.

Reply:
For some context to the DataStream team: We only plan on having Rollup Search Resolution happen when users query datastreams. Wherein underlying rollup indices will take precedence over their respective original indices, where the rollup index can answer the aggregation query. When users query rollup indices outside the scope of datastreams, it behaves just like a regular index so no new metadata associated is needed. This is coupled with DataStream.